### PR TITLE
Nordigen integration - account sync

### PR DESCRIPTION
**Pull request by fstybel**
This PR integrates the app with [Nordigen](https://nordigen.com/en/) to allow account sync with banks - balance, account info, and transfers.

Related backend PR: https://github.com/actualbudget/actual-server/pull/74

## Bank integrations

Unfortunately, Nordigen not provides one common interface for all banks' data we have to add write implementation for new banks. I wrote a simple [README file](https://github.com/actualbudget/actual-server/blob/087eff6cd4884ecb31eaa5c1ddbdc01a8aff31e4/app-nordigen/README.md) which describes it. But generally, we have to create a new class for the new bank and add implementation for three methods. 

I provided at the beginning implementation for two Poland Banks and for Nordigen Sandbox bank for tests. 

### Available banks

- ING PL
- MBANK Retail PL
- NORDIGEN Sandbox (for testing)

## Demo

https://user-images.githubusercontent.com/8612503/189112287-ac8f70ac-3e6f-4586-b6c2-99cda514db46.mov

## How to setup it up locally

1. Create a Nordigen account and generate API credentials
2. Put keys in the `config.json` file in the root of the backend repository

```
{
  "nordigen_secret_id": "xxxx",
  "nordigen_secret_key": "xxxxx"
}

```

3. Launch backend from the [PR](https://github.com/actualbudget/actual-server/pull/74) with Nordigen API credentials ` yarn start`
4. Launch frontend: `yarn start:browser`

### Backend

- [ ] Create a new endpoint that provides a list of the integrated banks
- [ ] Improve error handling (request limit exceeded, Bank service not available)

### Frontend 

- [ ] Add tests on the frontend side
- [ ] Fetch available bank integration from the backend instead of a hardcoded list
- [ ] Tweek the UI for new functionalities
- [ ] Add an option to select for how long period we want to establish access to our bank ( backend is ready for that)
- [ ] Add a progress bar to inform users about the process of importing data after when we add new accounts 
- [ ] **Add linking transactions between our accounts**
- [ ] Display somewhere when access to the account will expire and automatically ask for re-link of accounts

### CODE REVIEW

File: packages/loot-design/src/components/modals/NordigenExternalMsg.js
Comment (j-f1) [inline packages/loot-design/src/components/modals/NordigenExternalMsg.js:75]: Can you gate this provider to debug/preview versions somehow?

File: packages/loot-design/src/components/modals/NordigenExternalMsg.js
Comment (j-f1) [inline packages/loot-design/src/components/modals/NordigenExternalMsg.js:113]: This message should be updated when running in a browser to reflect that we’ll just open another tab. (same for plaid)

File: packages/loot-design/src/components/modals/NordigenExternalMsg.js
Comment (j-f1) [inline packages/loot-design/src/components/modals/NordigenExternalMsg.js:125]:
```suggestion
                  ? 'Waiting on Nordigen...'
```


(or something else giving more detail on why we’re waiting)

File: packages/loot-design/src/components/modals/NordigenExternalMsg.js
Comment (j-f1) [inline packages/loot-design/src/components/modals/NordigenExternalMsg.js:171]: This isn’t relevant if running in the browser.

File: packages/loot-design/src/components/modals/SelectLinkedAccounts.js
Comment (j-f1) [inline packages/loot-design/src/components/modals/SelectLinkedAccounts.js:97]: typo:

```suggestion
      // Try to auto-match accounts based on account_id or mask
```

**Review by j-f1 — COMMENTED** [review]
Some initial thoughts:

- It looks like you’ve added a top-level setting to the settings page. Would you be willing to put it behind Settings → Advanced → Experimental features for now, or are you confident enough to commit to it being supported as-is right now? (I’d like to lean towards not fully shipping it just yet, especially until we have tests & more banks supported)
- It seems like you’re removing the existing Plaid code from the frontend. Seeing how not all of our users are in Europe, would there be a way to have multiple sync services? This would defintiely require more UI so could be handled in a follow-up PR but I’m wary of deleting the existing Plaid code.
- The list of accounts is a dropdown, ideally it would be an Autocomplete like the payee/category pickers so that people can either scroll the list or type in the name of their bank
- Adding a synced account should only be allowed if you are connected to a server, and there should be an explanation of why the server is required in a message next to the disabled “Link bank account” button.
- [nit] can you remove `packages/mobile`? It looks like some deleted files have snuck back in.

**Comment by j-f1** [comment]
(That said, if the other maintainers are OK with this route, as long as all of these code paths are behind the experimental feature toggle I’m ok to merge soon and then iterate on labels/details later)

**Comment by MatissJanis** [comment]
👋 I love the work you've done so far to bring bank sync integration to Actual! Amazing to see what the community can build.

This feature has sparked some discussions within the management team and I wanted to also share it with you in case you don't follow the discussions on discord: https://discord.com/channels/937901803608096828/1027952080750661702/1072624203964092486

Long story short: we're putting __this feature__ to the vote. If the vote swings your way, then I'll do a more extensive review (even though Jed already made some very good points that should be addressed) and help get this released.

I'm very sorry that this we did not communicate this earlier to you. I thought this change would be much less controversial, so I take full responsibility. But at the end of the day we must respect the collective opinion of the community.

**Comment by fstybel** [comment]
Hey @MatissJanis 
Thanks for letting me know about it - I didn't know about the poll.  But I don't see any conclusion after it or it's still in progress? 

IMO we have some people who want this feature we should do it. Folks who don't want this can just not enable it.

**Comment by MatissJanis** [comment]
> Thanks for letting me know about it - I didn't know about the poll. But I don't see any conclusion after it or it's still in progress?

Still in progress. I'll keep it open until 19th Feb so people have enough time to share their thoughts and vote. Will post another message once the voting has concluded.

**Comment by rich-howell** [comment]
> Hey @MatissJanis Thanks for letting me know about it - I didn't know about the poll. But I don't see any conclusion after it or it's still in progress?
> 
> IMO we have some people who want this feature we should do it. Folks who don't want this can just not enable it.

The main reason this came about was because I raised it in a management meeting, I want to add here why I did that. 

Actual is a local first Budget application, the reason it is local first is that it puts you in control of your data and thus one of the main selling points is it is privacy focused which a lot of other tools are not.

People are going to counter my argument by saying they don't see why adding the option to allow people to choose if they share their data or not mutes what I am saying, the problem is it doesn't the fact the option is there means that Actual is no longer privacy focussed. 

Some of these providers openly say they are going to collect your data for various reasons, this means Actual is no longer privacy focussed, if you have the feature on or off is irrelevant, the option to allow third parties to potentially collect your data is there. 

I don't want to go into too much detail as I feel it is a conflict of interest but the above is why this got raised and why I wanted to raise it with the community before we add it to the code base.

File: packages/loot-core/src/server/accounts/sync.js
Comment (MatissJanis) [inline packages/loot-core/src/server/accounts/sync.js:531]: **❓ question:** ‏should we remove this?

File: packages/loot-core/src/server/main.js
Comment (MatissJanis) [inline packages/loot-core/src/server/main.js:1207]: **❓ question:** ‏do we need to check the feature flag here?

IMO if the flag is turned OFF (bank sync disabled), then `count` would always be `0`. So we never evaluate this expression.

**Review by MatissJanis — CHANGES_REQUESTED** [review]
A few small changes requested, but overall looks good. Pulled and tested locally too: works great!

Would you mind also taking a look at the build/lint issues in CICD?

We'll want to work on improving the UI/UX after it's merged (and before we release it as a first-party feature). But IMO the community will help with that once we get the current changes in master branch.

Great job! Lets push this over the finish line!

File: packages/loot-core/src/server/main.js
Comment (fstybel) [reply packages/loot-core/src/server/main.js:1207]: We should check it here but I don't know how to read the flag in this place... 
We need it because the `account-unlink` function is executed when the user closes the account. It'll fail in the following scenario - The user turned ON the flag and create a linked account. Later he switches the flag to OFF and he wants to close the account. Here the `count` returns more or equal to 1 and calls to nordigen endpoint.

**Comment by fstybel** [comment]
I have some issue with Netlify:
```
5:24:02 PM: ➤ YN0001: │ Error: While cloning /opt/build/repo/packages/loot-core/node_modules/cross-env -> /opt/build/repo/packages/desktop-client/node_modules/cross-env ENOENT: no such file or directory, scandir '/opt/build/repo/packages/loot-core/node_modules/cross-env'
```

**Comment by fstybel** [comment]
I'm not sure what we would like to do with the Plaid code. I can leave it as was previously but in a few places, I will have to remove/comment on it because this implementation is based on the Plaid code.

File: packages/loot-core/src/server/main.js
Comment (j-f1) [inline packages/loot-core/src/server/main.js:1075]: Currently this causes the call in the frontend to fail because it tries to destructure null.

```suggestion
  return { error: 'unauthorized' };
```

File: packages/loot-core/src/server/main.js
Comment (j-f1) [reply packages/loot-core/src/server/main.js:1207]: I think it’s fine to say that if you want to uncheck the feature flag you need to unlink all synced accounts or else you’ll run into issues.

**Comment by j-f1** [comment]
I'm seeing the Netlify issue too, not sure why it is happening.

File: packages/desktop-client/src/nordigen.js
Comment (j-f1) [reply packages/desktop-client/src/nordigen.js:13]: ^ otherwise the code crashes if the request fails.

**Comment by j-f1** [comment]
I’m getting an error on the line `err.message = 'Failed syncing account: ' + err.message;` because the `message` property is readonly.

**Comment by fstybel** [comment]
> I’m getting an error on the line `err.message = 'Failed syncing account: ' + err.message;` because the `message` property is readonly.

@j-f1 On my side all works very well. I tested it e2e a few minutes ago on the fresh budget. Did you follow instructions on how to run it locally from the description of this PR --> run separately backend from the backend PR, and run separately frontend from this PR.

**Comment by j-f1** [comment]
I ran it locally. I’m not seeing either this issue or the one I mentioned about the page staying open anymore. However I’m getting this error back from the API when I try to sync:

```js
 {
    date_to: {
      summary: "Date can't be in future",
      detail: "'2023-02-28' can't be greater than 2023-02-27. Specify correct date range"
    },
    status_code: 400
  }
```

**Comment by fstybel** [comment]
> I ran it locally. I’m not seeing either this issue or the one I mentioned about the page staying open anymore. However I’m getting this error back from the API when I try to sync:
> 
> ```js
>  {
>     date_to: {
>       summary: "Date can't be in future",
>       detail: "'2023-02-28' can't be greater than 2023-02-27. Specify correct date range"
>     },
>     status_code: 400
>   }
> ```

Oh.. I thought that I fixed it [here](https://github.com/actualbudget/actual/pull/457/files#diff-e8a250aeab2103544363c70f9ebdccaa705ea13d72703d69e30d032e02a2251fR188-R193). We have to pass `endDate` in UTC date. In which TZ are you now?

**Comment by j-f1** [comment]
I’m in US Eastern time

**Comment by fstybel** [comment]
> I’m in US Eastern time

@j-f1  Fixed

**Comment by j-f1** [comment]
I pushed a few changes to your branch to improve error handling and fix an issue with very large changes (like downloading 500 transactions at once). Confirmed that things seem to work OK in Safari!

**Review by j-f1 — APPROVED** [review]
LGTM! I have some things I’d like to tweak/improve but we can iterate on this more in future PRs.

**Comment by fstybel** [comment]
✅ Your changes look good! Thanks for your contribution ;)

**Comment by bjornbostrom** [comment]
Tested a local build with the Sandbox Integration and found it to work well. Also tested to create a new class for integrating the Swedish bank Handelsbanken, and it also went really smooth, and I got all my expected transactions automatically imported into Actual :) Great work!

**Review by MatissJanis — APPROVED** [review]
🎉 

I'll open up an issue where we can track the bugs after this is merged. IMO this is good to go out in the next release because it's under an experimental feature flag. We can get people testing it and reporting any issues. And then in the next-next release (hopefully) we'll have most of the issues addressed and we can offer this as a first-party feature.

**Comment by j-f1** [comment]
@fstybel Thank you so much for getting this over the line! I’m super excited to have bank integrations supported in Actual :)

**Pull request merged**
